deepwell/src/methods/page.rs

```diff
@@ -123,53 +227,114 @@ pub async fn page_links_to_missing_get(req: ApiRequest) -> ApiResponse {
     txn.commit().await?;
-
     Ok(body.into())
 }
 
-// TODO: remove separate endpoint, make part of revision changes
-pub async fn page_links_put(mut req: ApiRequest) -> ApiResponse {
+pub async fn page_links_external_from(req: ApiRequest) -> ApiResponse {
     let txn = req.database().begin().await?;
     let ctx = ServiceContext::new(&req, &txn);
-    let backlinks: Backlinks = req.body_json().await?;
 
     let site_id = req.param("site_id")?.parse()?;
     let reference = Reference::try_from(&req)?;
-    LinkService::update(&ctx, site_id, reference, &backlinks)
+    tide::log::info!(
+        "Getting external links from page {reference:?} in site ID {site_id}",
+    );
+
+    let page = PageService::get(&ctx, site_id, reference).await.to_api()?;
+    let output = LinkService::get_external_from(&ctx, page.page_id)
         .await
         .to_api()?;
 
+    let body = Body::from_json(&output)?;
     txn.commit().await?;
-
-    Ok(Response::new(StatusCode::NoContent))
+    Ok(body.into())
 }
 
-// TODO: remove separate endpoint, make part of revision changes
-pub async fn page_links_missing_put(mut req: ApiRequest) -> ApiResponse {
+pub async fn page_links_external_to(req: ApiRequest) -> ApiResponse {
     let txn = req.database().begin().await?;
     let ctx = ServiceContext::new(&req, &txn);
-    let backlinks: Backlinks = req.body_json().await?;
 
     let site_id = req.param("site_id")?.parse()?;
-    let slug = req.param("slug")?;
-    LinkService::update_missing(&ctx, site_id, slug, &backlinks)
+    let url = req.param("url")?;
+    tide::log::info!("Getting external links to URL {url} in site ID {site_id}");
+
+    let output = LinkService::get_external_to(&ctx, site_id, url)
         .await
         .to_api()?;
 
+    let body = Body::from_json(&output)?;
     txn.commit().await?;
+    Ok(body.into())
+}
 
-    Ok(Response::new(StatusCode::NoContent))
+async fn get_text(
+    ctx: &ServiceContext<'_>,
+    should_fetch: bool,
+    text_hash: &[u8],
+) -> Result<Option<String>> {
+    if !should_fetch {
+        return Ok(None);
+    }
+
+    let text = TextService::get(ctx, text_hash).await?;
+    Ok(Some(text))
 }
 
-// TODO: include current revision data too
-fn build_page_response(page: &PageModel, status: StatusCode) -> ApiResponse {
-    let body = Body::from_json(page)?;
+async fn build_page_response(
+    ctx: &ServiceContext<'_>,
+    page: &PageModel,
+    revision: &PageRevisionModel,
+    details: PageDetailsQuery,
+    status: StatusCode,
+) -> Result<Response> {
+    // Get category slug from ID
+    let category =
+        CategoryService::get(ctx, page.site_id, Reference::from(page.page_category_id))
+            .await?;
+
+    // Get text data, if needed
+    let wikitext = get_text(ctx, details.wikitext, &revision.wikitext_hash).await?;
+    let compiled_html =
+        get_text(ctx, details.compiled_html, &revision.compiled_hash).await?;
+
+    let output = GetPageOutput {
```

Comment:
I think this would be more elegant if it used a `From` trait? `GetPageOutput::From(page)` or whatever

Reply:
It actually originally was that, but at present it pulls from `PageModel`, `PageRevisionModel`, and two `Option<String>`s, so the `From` would be kind of messy (probably a tuple of all of them?)